extensions/standard-processors/processors/GetFile.cpp

```diff
@@ -158,76 +147,89 @@ void GetFile::onSchedule(core::ProcessContext *context, core::ProcessSessionFact
       performListing(request_);
       last_listing_time_.store(utils::timeutils::getTimeMillis());
     }
   }
 
-  const bool isDirEmptyAfterPoll = isListingEmpty();
-  logger_->log_debug("Is listing empty after polling directory %i", isDirEmptyAfterPoll);
-
-  if (!isDirEmptyAfterPoll) {
-    try {
-      std::queue<std::string> list;
-      pollListing(list, request_);
-      while (!list.empty()) {
-        std::string fileName = list.front();
-        list.pop();
-        logger_->log_info("GetFile process %s", fileName);
-        auto flowFile = session->create();
-        if (flowFile == nullptr)
-          return;
-        std::size_t found = fileName.find_last_of("/\\");
-        std::string path = fileName.substr(0, found);
-        std::string name = fileName.substr(found + 1);
-        flowFile->setAttribute(core::SpecialFlowAttribute::FILENAME, name);
-        flowFile->setAttribute(core::SpecialFlowAttribute::PATH, path);
-        flowFile->addAttribute(core::SpecialFlowAttribute::ABSOLUTE_PATH, fileName);
-        session->import(fileName, flowFile, request_.keepSourceFile);
-        session->transfer(flowFile, Success);
+  const bool is_dir_empty_after_poll = isListingEmpty();
+  logger_->log_debug("Listing is %s after polling directory", is_dir_empty_after_poll ? "empty" : "not empty");
+  if (is_dir_empty_after_poll) {
+    yield();
+    return;
+  }
+
+  std::queue<std::string> list_of_file_names = pollListing(request_.batchSize);
+  while (!list_of_file_names.empty()) {
+    std::string file_name = list_of_file_names.front();
+    list_of_file_names.pop();
+    getSingleFile(*session, file_name);
+  }
+}
+
+void GetFile::getSingleFile(core::ProcessSession& session, const std::string& file_name) const {
+  logger_->log_info("GetFile process %s", file_name);
+  auto flow_file = session.create();
+  gsl_Expects(flow_file);
+  std::string path;
+  std::string name;
+  std::tie(path, name) = utils::file::split_path(file_name);
+  flow_file->setAttribute(core::SpecialFlowAttribute::FILENAME, name);
+  flow_file->setAttribute(core::SpecialFlowAttribute::PATH, path);
+  flow_file->addAttribute(core::SpecialFlowAttribute::ABSOLUTE_PATH, file_name);
+
+  try {
+    utils::FileReaderCallback file_reader_callback{file_name};
+    session.write(flow_file, &file_reader_callback);
+    session.transfer(flow_file, Success);
+    if (!request_.keepSourceFile) {
+      auto remove_status = remove(file_name.c_str());
+      if (remove_status != 0) {
+        logger_->log_error("GetFile could not delete file '%s', error %d: %s", file_name, errno, strerror(errno));
       }
-    } catch (std::exception &exception) {
-      logger_->log_debug("GetFile Caught Exception %s", exception.what());
-      throw;
-    } catch (...) {
-      throw;
     }
+  } catch (const std::exception& exception) {
+    logger_->log_error("GetFile caught exception while processing file '%s': %s", file_name, exception.what());
+    flow_file->setDeleted(true);
```

Comment:
Do we need this? 
I would expect session->rollback() later to clean up all the "mess" we made in an interrupted onTrigger

Reply:
We don't want to roll back if reading one file failed but another file succeeded; in this case, we want to ignore the failed file but process the successful one.

My understanding is that every flow file we create with `session->create()` has to either be transferred to an output relationship or marked as deleted, otherwise `routeFlowFile(flow_file)`, and therefore `session->commit()`, will throw.

Reply:
Okay, I need more coffee, didn't realise that this is a different scenario. :)